### ISSUE TITLE
convert xgboost model to pmml model failed

### ISSUE THREAD

**Issue by xixi2715** [jpmml/jpmml-xgboost#23]
I used the following code to save a xgboost model
```
import xgboost as xgb
import numpy as np

dtrain = xgb.DMatrix('/Users/yinfengzhao/PycharmProjects/xgboost/demo/data/agaricus.txt.train')

param = {'max_depth':3, 'eta':1, 'silent':1, 'objective':'binary:logistic' }
num_round = 50
bst = xgb.train(param, dtrain, num_round)
bst.save_model('demo.model')
```

the dataset used is from xgboost project, and I find a featmap.txt there, I used the folowing code to convert the xgboost model to pmml model:
```
java -jar converter-executable-1.2-SNAPSHOT.jar --model-input demo.model --fmap-input featmap.txt --target-name f0 --pmml-output xgboost_demo.pmml
```

but I got a error like this:
```
Exception in thread "main" java.lang.IndexOutOfBoundsException: Index: 126, Size: 126
	at java.util.ArrayList.rangeCheck(ArrayList.java:653)
	at java.util.ArrayList.get(ArrayList.java:429)
	at org.jpmml.converter.Schema.getFeature(Schema.java:81)
	at org.jpmml.xgboost.RegTree.encodeNode(RegTree.java:113)
	at org.jpmml.xgboost.RegTree.encodeNode(RegTree.java:123)
	at org.jpmml.xgboost.RegTree.encodeTreeModel(RegTree.java:95)
	at org.jpmml.xgboost.ObjFunction.createMiningModel(ObjFunction.java:65)
	at org.jpmml.xgboost.BinomialLogisticRegression.encodeMiningModel(BinomialLogisticRegression.java:43)
	at org.jpmml.xgboost.GBTree.encodeMiningModel(GBTree.java:77)
	at org.jpmml.xgboost.Learner.encodeMiningModel(Learner.java:151)
	at org.jpmml.xgboost.Learner.encodePMML(Learner.java:140)
	at org.jpmml.xgboost.Main.run(Main.java:148)
	at org.jpmml.xgboost.Main.main(Main.java:128)
```
How did it happen? Can anyone help me?
[featmap.txt](https://github.com/jpmml/jpmml-xgboost/files/1770982/featmap.txt)

**Comment by vruusmann** [comment]
The most likely explanation is that `agaricus.txt.train` and `featmap.txt` are out of sync. The trained XGBoost model contains an instruction "split the node based on feature at position 126 of the feature map", but the last index of the feature map is 125 (assuming zero-based indexing).

If you're working with Python, then it's advisable to use XGBoost via its Scikit-Learn wrapper (eg. the `XGBClassifier` class). The SkLearn2PMML/JPMML-SkLearn converter stack is able to construct up-to-date feature map description automatically by observing pipeline steps.

**Comment by xixi2715** [comment]
@vruusmann ，Thanks for your advise，but I still feel confused.
 the link of data  `agaricus.txt.train` is this: https://github.com/dmlc/xgboost/blob/master/demo/data/agaricus.txt.train. The data contains 126 features, but after using this code:
```
dtrain = xgb.DMatrix('/Users/yinfengzhao/PycharmProjects/xgboost/demo/data/agaricus.txt.train')
print dtrain.feature_names
```
it print 127 features columns. I don't understand why. Do you know?

**Comment by vruusmann** [comment]
> the link of data agaricus.txt.train is this: 

This is a data file in LibSVM data format. The first column is a label column (`0` or `1` value), which is followed by 126 feature columns (sparse representation).

I don't know if it's permitted to read LibSVM file into a `DMatrix` object like that. Perhaps you should load it first into some temporary object, remove the label column, and construct the real feature columns `DMatrix` object based on the remaining data.

Overall, the `DMatrix` class has rather poor end user API in my opinion. Like mentioned in my previous comment, try to approach XGBoost library using high-level frameworks such as Scikit-Learn or R, which will automatically take care of feature maps and `DMatrix` objects.

**Comment by xixi2715** [comment]
@vruusmann OK, I will try to use Scikit-Learn. Thanks for your advise. You are very thoughtful.
